### ISSUE TITLE
Exceptions on HEAD and OPTIONS requests to the IDP

### ISSUE THREAD

**Issue by htto** [simplesamlphp/simplesamlphp#1400]
**Describe the bug**
When a client (noticed Firefox, Chrome, Edge and Word in the logs) sends an OPTIONS or HEAD request, SSP throws an unhandled Exception and the server then returns status 500. In a quick look I couldn't find methods OPTIONS/HEAD in the source nor an existing issue.

**To Reproduce**
1. I've seen the exceptions on a real server, but wasn't able to trigger them myself in a standard login-ish way
2. But one probably can come close enough with curl
3. curl -Iv -L https://site/some-sp-protected-app
4. The trace looks pretty similar

**Expected behavior**
Not quite sure what to expect in a standardized way, but not running into 500 would be a start.

**Screenshots or logs**
Trace from a production server with site/path redacted
```
simplesamlphp-idp WARNING [cfc72b4cab] Request method: 'OPTIONS', referer: https://site/
simplesamlphp-idp WARNING [cfc72b4cab] GET parameters: 'SAMLRequest', 'RelayState', referer: https://site/
simplesamlphp-idp ERR [cfc72b4cab] SimpleSAML\\Error\\Error: UNHANDLEDEXCEPTION, referer: https://site/
simplesamlphp-idp ERR [cfc72b4cab] Backtrace:, referer: https://site/
simplesamlphp-idp ERR [cfc72b4cab] 1 simplesamlphp/simplesamlphp/www/_include.php:17 (SimpleSAML_exception_handler), referer: https://site/
simplesamlphp-idp ERR [cfc72b4cab] 0 [builtin] (N/A), referer: https://site/
simplesamlphp-idp ERR [cfc72b4cab] Caused by: Exception: Unable to find the SAML 2 binding used for this request., referer: https://site/
simplesamlphp-idp ERR [cfc72b4cab] Backtrace:, referer: https://site/
simplesamlphp-idp ERR [cfc72b4cab] 2 simplesamlphp/saml2/src/SAML2/Binding.php:107 (SAML2\\Binding::getCurrentBinding), referer: https://site/
simplesamlphp-idp ERR [cfc72b4cab] 1 simplesamlphp/simplesamlphp/modules/saml/lib/IdP/SAML2.php:357 (SimpleSAML\\Module\\saml\\IdP\\SAML2::receiveAuthnRequest), referer: https://site/
simplesamlphp-idp ERR [cfc72b4cab] 0 simplesamlphp/simplesamlphp/www/saml2/idp/SSOService.php:21 (N/A), referer: https://site/
simplesamlphp-idp ERR [cfc72b4cab] Error report with id 45d1887a generated., referer: https://site/
```

**Additional context**
simplesamlphp/saml2                            v4.1.10
simplesamlphp/simplesamlphp              v1.18.8

**Comment by tvdijen** [comment]
Thanks @htto !  I guess we could put up a more friendly message, but I do consider this an edge-case scenario.

**Comment by thijskh** [comment]
I would be in favour of handling it better since quite some browsers use 'options' requests as a preloading measure. While it does not hurt hard to give an error on it, it pollutes the logs, hiding real exceptions. So best is to return a method not implemented. Probably as easy as adding the allowed http methods to the routes in question. Then others will be rejected with a 405.

**Comment by htto** [comment]
Hey. Yeah, it's not a major case, but we get it on almost a daily basis for a handful of requests, though I dunno how they manage to trigger the preflights, no even sure if they see an error at all or if it's silently ignored.

Anyway, I'm all in for some less rigid approach, even if its a WARN instead of ERR/500. I'm not quite sure how to properly respond to HEAD/OPTIONS if unsupported. According to [1] HEAD isn't allowed to be responded to with 405/501. Seems like it must be supported?

[1] https://developer.mozilla.org/en-US/docs/Web/HTTP/Status

**Comment by thijskh** [comment]
That's a bit theoretical, we forbid OPTIONS/PROPFIND/HEAD in our Apache vhost (for paths that will not be able to return anything sensible for it) and nothing breaks, it does prevent spurious and useless error logs.

**Comment by tvdijen** [comment]
Yes, I agree with @thijskh that we should take care of this with Symfony routing.. However I also agree with @htto that a HEAD should never result in a 405 error.. I know many people disable it in the web server, but it should just be treated like a GET, without the response content

**Comment by cicnavi** [comment]
Since HTTP OPTIONS can be used, among other things, for checking for the allowed HTTP methods to be sent in the real request, is there some reason to not answer the supported HTTP methods in SSP (GET and POST)? 

https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods/OPTIONS

**Comment by tvdijen** [comment]
Well, I'd say it's safe, but if you're really paranoid it could be used to figure out the capabilities of the webserver (and by that, possible attack vectors).. Most vulnerability-scanners will report it

**Comment by pokej6** [comment]
Supporting HEAD and OPTIONS doesn't really hurt anything but if the decision is to not support them, then perhaps it's a good idea to make use of SAML StatusCodes: [See 3.2.2.2 Element <StatusCode>](https://www.oasis-open.org/committees/download.php/56777/sstc-saml-core-errata-2.0-wd-07-diff.pdf) potentially in addition to HTTP response codes. In this way something that speaks SAML may be able to gracefully handle the error. This is similar to how JSONRPC errors work.

**Comment by tvdijen** [comment]
I strongly disagree.. We shouldn't mix HTTP and SAML for that matter. There's no way in hell a SAML requester is using any method other than GET or POST, so it makes zero sense to reply with a SAML error when HEAD or OPTIONS is being used.

**Comment by cicnavi** [comment]
When implementing or disabling HEAD / OPTIONS requests generally, also note that SSP has modules which might be dependent on the mentioned requests, like for example module 'oidc' which provides authN using OpenID Connect protocol. In OIDC, it is common to have JavaScript clients, meaning clients running solely in web browsers. From the frontend perspective, this authn is a cross-origin scenario, and depending on the OIDC authN flow, the browser might issue a preflight request to check if it is allowed to make a real cross-origin request (note that the browser does it automatically, not the OIDC client). If the preflight request would be marked as 405 before it reaches the 'oidc' module, it would break the OIDC authN. So, if disabling OPTIONS, it should be done on the SAML module level (sorry for the lack of better designation), and let other modules handle it themselves.

From https://developer.mozilla.org/en-US/docs/Glossary/Preflight_request
> A preflight request is automatically issued by a browser and in normal cases, front-end developers don't need to craft such requests themselves. It appears when request is qualified as "to be preflighted" and omitted for simple requests.

**Comment by tvdijen** [comment]
But we're not SimpleOIDCPHP...  so it should not be disabled on the saml-module, but the other way around.. If a module requires this, it can be enabled in the modules routing config.

**Comment by cicnavi** [comment]
Sure, I understand and I agree. Just wanted to point it out, so that it is possible to use those requests in modules, in whatever way possible...

**Comment by tvdijen** [comment]
Yes, you can set `methods:  [OPTIONS, HEAD, GET]` in a module's routing.yml file, so that shouldn't be a problem.

**Comment by pokej6** [comment]
> There's no way in hell a SAML requester is using any method other than GET or POST

Well this doesn't seem to be the case otherwise this issue wouldn't have been filed. It does make sense that you wouldn't response to a HEAD request with a response body, though.

**Comment by tvdijen** [comment]
As you've said yourself you haven't been able to trigger it in a standard situation.. Might as well have been some vulnerability scanner

**Comment by pokej6** [comment]
> As you've said yourself you haven't been able to trigger it in a standard situation

I don't recall saying that. A "SAML Requester" is generally just a browser (browser profile for SAML is very common). Browsers do make HEAD and OPTIONS requests.

I've seen a SAML requester make a HEAD request before. When you click on links in Microsoft Office products they will follow the url and redirects in order to check the validity of the link. This can result in a HEAD request being made to a SAML IdP. Office products such as Excel use your IE/Edge cookie store so it's possible you are seamlessly logged in unless you happen to error out on the HEAD request.

**Comment by htto** [comment]
OK, not quite sure if it's of any help, but it seems one can trigger preflights for AJAX when using non-simple requests (e.g. by adding Header 'Authorization'), not necessarily cross-domain. I could think of some corporate site (with some sort of HTTP Authorization) including our site for ease of use. 

Something like
```
 $ cat ssp_test.html 
<html>
<head>
<script>
var x = new XMLHttpRequest();
x.open('GET', 'http://localhost/ssp_test_target.php');
x.setRequestHeader('Authorization', 'somevalue');
x.send();
</script>
</head>
</html>
```
```
 $ cat ssp_test_target.php 
<?php
require __DIR__.'/vendor/autoload.php';

Simple->requireAuth();
```
invoked with http://localhost/ssp_test.html, can replicate the issue.

Though I'm not quite sure how the users manage to get there with Word 2014 as an user agent.

**Comment by tvdijen** [comment]
Closed in a7600ce477632aca33eb206be7c41056ce808e14

The proper solution will be implemented in 2.0 using Symfony routing files.
